src/fortests/SetupModel.cpp

```diff
@@ -65,70 +69,59 @@ SetupModel::SetupModel(const double& rHertz,
-        CellMLLoader loader(cellml_file, *mpHandler, options);
+        CellMLLoader loader(cellml_file, *mpHandler, {});
         mpModel = loader.LoadCvodeCell();
-    }
-    else // Using a hardcoded model
-    {
-        if (modelIndex == UNSIGNED_UNSET)
-        {
-            if (!CommandLineArguments::Instance()->OptionExists("--model"))
-            {
-                EXCEPTION("Argument \"--model <index>\" is required");
-            }
-            modelIndex = CommandLineArguments::Instance()->GetUnsignedCorrespondingToOption("--model");
+
+    }else{ // we have been given a model name or number
+        if(CommandLineArguments::Instance()->OptionExists("--cellml")){
+            EXCEPTION("Invalid file given with --cellml argument: " + modelName);
         }
-        switch (modelIndex)
-        {
-        case 1u:
-            // This one is from the cellml project - more metadata.
-            mpModel.reset(new Cellshannon_wang_puglisi_weber_bers_2004FromCellMLCvode(p_solver, p_stimulus));
-            // This one is from the Chaste source
-            //mpModel.reset(new CellShannon2004FromCellMLCvode(p_solver, p_stimulus));
-            break;
-        case 2u:
-            mpModel.reset(new Cellten_tusscher_model_2006_epiFromCellMLCvode(p_solver, p_stimulus));
-            break;
-        case 3u:
-            mpModel.reset(new Cellmahajan_shiferaw_2008FromCellMLCvode(p_solver, p_stimulus));
-            break;
-        case 4u:
-            mpModel.reset(new Cellhund_rudy_2004FromCellMLCvode(p_solver, p_stimulus));
-            // Hund Rudy doesn't play well with the use of an Analyic Jacobian, see Cooper, Spiteri, Mirams, 2015 paper
-            mpModel->ForceUseOfNumericalJacobian(true);
-            break;
-        case 5u:
-            mpModel.reset(new Cellgrandi_pasqualini_bers_2010_ssFromCellMLCvode(p_solver, p_stimulus));
-            break;
-        case 6u:
-            mpModel.reset(new Cellohara_rudy_2011_endoFromCellMLCvode(p_solver, p_stimulus));
-            break;
-        case 7u:
-            mpModel.reset(new Cellpaci_hyttinen_aaltosetala_severi_ventricularVersionFromCellMLCvode(p_solver, p_stimulus));
-            break;
-        case 8u:
-            mpModel.reset(new Cellohara_rudy_cipa_v1_2017FromCellMLCvode(p_solver, p_stimulus));
-            break;
-        default:
-            EXCEPTION("No model matches this index");
+        
+        // Check if we have been given an index that can be mapped to a model name
+        auto mapIterator = SetupModel::modelMapping.find(modelName);
+        if(mapIterator != SetupModel::modelMapping.end()){
+            modelName = mapIterator->second;
         }
+
+        // Create model using factory
+        mpModel.reset((AbstractCvodeCell*)ModelFactory::Create(modelName , "AnalyticCvode", p_solver, p_stimulus));
+
+        if(mpModel == nullptr){  // throw an error if the model isn't found
+            EXCEPTION("No model matches this index: " + modelName);
```

Comment:
Is this exception tested? I'd expect this logic perhaps to be in an `else` on line 113-114 so that we don't try and create a model that doesn't exist?

Reply:
You're right. I've added an error to the creation method when creating a non-existing model, and an Exists mehod to check if a model/type combo can be created. Also updated tests
see `TestModelFactory.hpp` `TestSetupModelCommandLineWrongIndex` and `TestWrongModelOrType`